### PR TITLE
Implement replacement for post-commit validation

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
PostCommitValidation is meant as a replacement for ActiveLedgerState and
ActiveLedgerStateManager for JdbcLedgerDao so that the old table can eventually
be retired and post-commit validation can happen on the new participant schema.

This commit simply introduces the implementation and tests it in isolation.

Integration with the existing system will come in a separate contribution.

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        data.lookupMaximumLedgerTime(
+          transaction.fold(Set.empty[ContractId]) {
+            case (ids, (_, e: Exercise)) if !divulged(e.targetCoid) => ids + e.targetCoid
+            case (ids, (_, f: Fetch)) if !divulged(f.coid) => ids + f.coid
+            case (ids, (_, l: LookupByKey)) => l.result.filterNot(divulged).fold(ids)(ids + _)
+            case (ids, _) => ids
+          }
+        )
+
+      // Query the committed contracts store for the maximum ledger effective time of the
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala:94]: Should this comment go above line 85? Looks like the querying is happening there.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+  /**
+    * Represents the state of an ongoing validation.
+    * It must be carried over as the transaction is
+    * validated one node at a time in pre-order
+    * traversal for this to make sense.
+    *
+    * @param errors Accumulates errors retrieved so far while validating
+    * @param contracts All contracts created as part of the current transaction
+    * @param removed Ensures indexed transactions are not referred to if they are removed in the current transaction
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala:152]: Should `indexed transactions` rather be `indexed contracts` or `indexed keys`?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala:94]: Yes, some code moved and this is a leftover, thanks.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala:152]: Absolutely, thanks!

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala]: Addressed by https://github.com/digital-asset/daml/pull/5737/commits/66a98902caf5384b1d9f95a612b06ccdb7e3f86b

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala]: Addressed by https://github.com/digital-asset/daml/pull/5737/commits/83d123331c49e94ed8209fd263cde54edac3957a

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    private def collectReferredContracts(
+        transaction: Transaction,
+        divulged: Set[ContractId],
+    ): Set[ContractId] = {
+      val (createdInTransaction, referred) =
+        transaction.fold((Set.empty[ContractId], Set.empty[ContractId])) {
+          case ((created, ids), (_, c: Create)) =>
+            (created + c.coid, ids)
+          case ((created, ids), (_, e: Exercise)) if !divulged(e.targetCoid) =>
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala:118]: Addressed by https://github.com/digital-asset/daml/pull/5774/commits/012d19624820ee6c898c3be68401bd948a67ceda in #5774.

**Pull request merged**